In [ ]:
import pandas as pd
import numpy as np
import torch
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    TrainingArguments,
    Trainer,
    DataCollatorForLanguageModeling
)
from datasets import Dataset
from sklearn.metrics import accuracy_score, precision_recall_fscore_support, roc_auc_score
from sklearn.model_selection import train_test_split
from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training
import time
import json
import random
from tqdm.auto import tqdm


In [ ]:
# Доли пропущенных признаков для экспериментов
MISSING_RATES = [0.0, 0.2, 0.5, 0.9]

# Настройка промпта для Blood
prompt_config = {
        "task": "Predict whether a person donated blood",
        "pos_label": "yes",
        "neg_label": "no",
        "entity": "Donor",
        "question": "Did this person donate blood?"
}

FEATURE_MAPPINGS = {
        'V1': 'Recency',
        'V2': 'Frequency',
        'V3': 'Monetary',
        'V4': 'Time',
        'Class': 'Donated blood'
}

openml_id = 1464

base_output_dir = "/content/drive/MyDrive/finetuned_qwen_missing_Blood"

# 1. Загрузка данных

In [ ]:
import pandas as pd
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

def load_dataset(openml_id=1464):
    dataset = fetch_openml(data_id=openml_id, as_frame=True, parser='auto')
    X = dataset.data
    y = dataset.target

    df = X.copy()
    feature_names = X.columns.to_list()

    df = df.rename(columns=FEATURE_MAPPINGS)
    feature_names = list(FEATURE_MAPPINGS.values())[:-1]
    target_name = list(FEATURE_MAPPINGS.values())[-1]
    df[target_name] = y

    # Преобразование целевой переменной в бинарный формат
    if y.dtype == 'object' or y.dtype.name == 'category':
        le = LabelEncoder()
        df[target_name] = le.fit_transform(df[target_name])

    return df, feature_names, target_name


def split_dataset(df, target_name, test_size=0.2, val_size=0.25, seed=42):

    # Разделение на train/val/test (60/20/20)
    train_val_df, test_df = train_test_split(
        df,
        test_size=test_size,
        random_state=seed,
        stratify=df[target_name]
    )

    train_df, val_df = train_test_split(
        train_val_df,
        test_size=val_size,
        random_state=seed,
        stratify=train_val_df[target_name]
    )

    return train_df, val_df, test_df

df, feature_names, target_name = load_dataset(openml_id)
train_df, val_df, test_df = split_dataset(df, target_name)

df.head(5)

,Recency,Frequency,Monetary,Time,Donated blood
0,2,50,12500,98,1
1,0,13,3250,28,1
2,1,16,4000,35,1
3,2,20,5000,45,1
4,1,24,6000,77,0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 748 entries, 0 to 747
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   Recency        748 non-null    int64
 1   Frequency      748 non-null    int64
 2   Monetary       748 non-null    int64
 3   Time           748 non-null    int64
 4   Donated blood  748 non-null    int64
dtypes: int64(5)
memory usage: 29.3 KB


In [ ]:
for name, df in [("train", train_df), ("val", val_df), ("test", test_df)]:
    counts = df[target_name].value_counts()
    pcts   = df[target_name].value_counts(normalize=True) * 100
    print(f"\n{name} (всего: {len(df)}):")
    print(f"  {prompt_config['neg_label']}: {counts[0]:5d} — {pcts[0]:.1f}%")
    print(f"  {prompt_config['pos_label']}: {counts[1]:5d} — {pcts[1]:.1f}%")


train (всего: 448):
  no:   342 — 76.3%
  yes:   106 — 23.7%

val (всего: 150):
  no:   114 — 76.0%
  yes:    36 — 24.0%

test (всего: 150):
  no:   114 — 76.0%
  yes:    36 — 24.0%


# 2. Вспомогательные функции

In [ ]:
def row_to_text_template_with_missing(row, feature_names, target_name, missing_rate=0.0, seed=None):

    rng = np.random.default_rng(seed)

    # Определяем, какие признаки пропустить
    n_drop = int(len(feature_names) * missing_rate)
    dropped = set(rng.choice(feature_names, size=n_drop, replace=False)) if n_drop > 0 else set()

    template_parts = []

    for feature in feature_names:
        if feature in dropped:
            continue  # пропускаем признак

        value = row[feature]

        if isinstance(value, (int, np.integer)):
            phrase = f"The value of {feature} is {value}."
        elif isinstance(value, (float, np.floating)):
            phrase = f"The value of {feature} is {value:.2f}."
        else:
            phrase = f"The category of {feature} is {value}."

        template_parts.append(phrase)

    text = " ".join(template_parts)

    return text

# Тест
print(row_to_text_template_with_missing(train_df.iloc[0], feature_names, target_name))
train_df.head(1)

The value of Recency is 9. The value of Frequency is 9. The value of Monetary is 2250. The value of Time is 45.


,Recency,Frequency,Monetary,Time,Donated blood
199,9,9,2250,45,0


In [ ]:
def parse_prediction(response, prompt_config):
    response = response.lower().strip().rstrip('.,!? ')
    pos, neg = prompt_config['pos_label'].lower(), prompt_config['neg_label'].lower()

    # Точное совпадение
    if response == pos: return 1
    if response == neg: return 0

    # Начинается с метки
    if response.startswith(pos): return 1
    if response.startswith(neg): return 0

    # Содержит метку как отдельное слово
    words = response.split()
    if pos in words: return 1
    if neg in words: return 0

    # Не удалось распознать
    print(f"Warning: Could not parse '{response}' (expected '{prompt_config['pos_label']}' or '{prompt_config['neg_label']}')")
    return 0



response = prompt_config['pos_label']
pred = parse_prediction(response, prompt_config)
print(f"Response: '{response}'\nPrediction: {pred}\n")

response = "hi"
pred = parse_prediction(response, prompt_config)
print(f"Response: '{response}'\nPrediction: {pred}")

Response: 'yes'
Prediction: 1

Response: 'hi'
Prediction: 0


In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
# Вычисление метрик качества
def compute_metrics(y_true, y_pred, y_prob=None):
    roc = roc_auc_score(y_true, y_prob if y_prob is not None else y_pred)
    return {
        "ROC-AUC": roc_auc_score(y_true, y_prob if y_prob is not None else y_pred),
        "F1":       f1_score(y_true, y_pred, zero_division=0),
        "Accuracy": accuracy_score(y_true, y_pred),
        "Precision": precision_score(y_true, y_pred, zero_division=0),
        "Recall":   recall_score(y_true, y_pred, zero_division=0),
    }

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.utils import resample

def bootstrap_metrics(y_true, y_pred, y_prob=None, n_iter=1000):
    scores = []

    for i in range(n_iter):
        # Bootstrap выборка
        if y_prob is not None:
            y_true_boot, y_pred_boot, y_prob_boot = resample(
                y_true, y_pred, y_prob, random_state=i+1
            )
        else:
            y_true_boot, y_pred_boot = resample(
                y_true, y_pred, random_state=i+1
            )
            y_prob_boot = None

        try:
            # Вычисление метрик
            if y_prob_boot is not None:
                auc = roc_auc_score(y_true_boot, y_prob_boot)
            else:
                auc = 0.0

            f1 = f1_score(y_true_boot, y_pred_boot, zero_division=0)
            pr = precision_score(y_true_boot, y_pred_boot, zero_division=0)
            rc = recall_score(y_true_boot, y_pred_boot, zero_division=0)

            acc = accuracy_score(y_true_boot, y_pred_boot)
            scores.append((auc, f1, acc, pr, rc))

        except ValueError:
            # Пропускание сэмплов, где не представлены все классы
            continue

    scores = np.asarray(scores)
    means, stds = scores.mean(0), scores.std(0, ddof=1)
    names = ["ROC-AUC", "F1", "Accuracy", "Precision", "Recall"]

    return {n: f"{m:.4f}±{s:.4f}" for n, m, s in zip(names, means, stds)}

# Тестирование функции
sample_row = test_df.iloc[0]
sample_text = row_to_text_template_with_missing(sample_row, feature_names, target_name)
print(f"\nПример текста:\n{sample_text[:300]}")


Пример текста:
The value of Recency is 4. The value of Frequency is 16. The value of Monetary is 4000. The value of Time is 38.


In [ ]:
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig

#Загрузка модели Qwen 3.0 4B Instruct
model_name = "Qwen/Qwen3-4B-Instruct-2507"

device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Using device: {device}")

# Загрузка токенизаторов
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Загрузка модель с 16-битной квантизацией и автоматическим распределением по устройствам
base_model = AutoModelForCausalLM.from_pretrained(
    model_name,
    dtype=torch.bfloat16,
    device_map=device, # Explicitly force all model layers to the detected device (GPU if available)
    attn_implementation="sdpa"
)

if torch.cuda.is_available():
    print(f"Использовано памяти: {torch.cuda.memory_allocated() / 1e9:.2f} GB")

Using device: cuda


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Loading weights:   0%|          | 0/398 [00:00<?, ?it/s]

Использовано памяти: 8.04 GB


In [ ]:
def create_prompt_missing(row, feature_names, target_name, prompt_config, tokenizer, missing_rate=0.0, seed=None):

    system_prompt = (
        f"You are a classifier. {prompt_config['task']}: "
        f"'{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'. "
        f"Answer with only one word: '{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'."
    )

    input_text = row_to_text_template_with_missing(row, feature_names, target_name, missing_rate, seed)

    messages = [
        {"role": "system", "content": system_prompt},
        {"role": "user",
         "content": f"{prompt_config['entity']} information: {input_text}\n{prompt_config['question']}"},
    ]

    return tokenizer.apply_chat_template(
        messages,
        tokenize=False,
        add_generation_prompt=True # добавление role assistant
    )

In [ ]:
import torch.nn.functional as F

def predict_single_with_prob(prompt, prompt_config, model, tokenizer, device, max_new_tokens=5):

    model_inputs = tokenizer([prompt], return_tensors="pt").to(device)

    with torch.no_grad():
        outputs = model.generate(
            model_inputs.input_ids,
            max_new_tokens=max_new_tokens,
            do_sample=False,
            pad_token_id=tokenizer.eos_token_id,
            output_scores=True,
            return_dict_in_generate=True
        )

    # Декодирование текста
    generated_ids = outputs.sequences[0][model_inputs.input_ids.shape[1]:]
    response = tokenizer.decode(generated_ids, skip_special_tokens=True).strip().lower()

    # Извлечение вероятностей для меток из prompt_config
    first_token_logits = outputs.scores[0][0]

    pos_id = tokenizer.encode(prompt_config['pos_label'], add_special_tokens=False)[0]
    neg_id = tokenizer.encode(prompt_config['neg_label'], add_special_tokens=False)[0]

    pos_logit = first_token_logits[pos_id]
    neg_logit = first_token_logits[neg_id]

    probs = F.softmax(torch.stack([pos_logit, neg_logit]), dim=0)
    prob_pos = probs[0].item()

    return response, prob_pos

# Тест
test_prompt = create_prompt_missing(test_df.iloc[0], feature_names, target_name, prompt_config, tokenizer)
response, prob = predict_single_with_prob(test_prompt, prompt_config, base_model, tokenizer, device)
print(f"Response: '{response}', Probability: {prob:.4f}")

The following generation flags are not valid and may be ignored: ['temperature', 'top_p', 'top_k']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Response: 'yes', Probability: 0.9975


# 3. Fine-tuning с LoRA

## Подготовка данных для Fine-tuning

In [ ]:
def balance_dataset(train_df, target_name, prompt_config, seed=42):
    df_0 = train_df[train_df[target_name] == 0]
    df_1 = train_df[train_df[target_name] == 1]

    df_majority = df_0 if len(df_0) > len(df_1) else df_1
    df_minority = df_1 if len(df_0) > len(df_1) else df_0

    print(f"До балансировки:")
    print(f"{prompt_config['neg_label']}:  {len(df_majority)}")
    print(f"{prompt_config['pos_label']}: {len(df_minority)}")

    df_minority_up = resample(df_minority, replace=True,
                              n_samples=len(df_majority), random_state=seed)

    train_df_balanced = pd.concat([df_majority, df_minority_up])
    train_df_balanced = train_df_balanced.sample(frac=1, random_state=seed).reset_index(drop=True)

    print(f"\nПосле балансировки:")
    print(train_df_balanced[target_name].value_counts())

    return train_df_balanced

train_df_balanced = balance_dataset(train_df, target_name, prompt_config)

До балансировки:
no:  342
yes: 106

После балансировки:
Donated blood
1    342
0    342
Name: count, dtype: int64


In [ ]:
# Создание текстов для обучения

def create_training_dataset(df, feature_names, target_name, prompt_config, tokenizer, missing_rate=0.0):
    system_prompt = (
        f"You are a classifier. {prompt_config['task']}: "
        f"'{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'. "
        f"Answer with only one word: '{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'."
    )
    texts = []

    for idx, (_, row) in enumerate(tqdm(df.iterrows(), total=len(df), desc=f"Dataset (missing={missing_rate:.0%})")):
        input_text = row_to_text_template_with_missing(row, feature_names, target_name, missing_rate, idx)
        target = prompt_config['pos_label'] if row[target_name] == 1 else prompt_config['neg_label']

        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": f"{prompt_config['entity']} information: {input_text}\n{prompt_config['question']}"},
            {"role": "assistant", "content": target}
        ]

        text = tokenizer.apply_chat_template(
            messages,
            tokenize=False,
            add_generation_prompt=False
        )
        texts.append(text)

    return texts


In [ ]:
def finetune_model(missing_rate, train_df_balanced, feature_names, target_name,
                   prompt_config, tokenizer, base_model, device,
                   num_epochs=3, batch_size=16):
    """
    Полный цикл fine-tuning для заданного missing_rate.
    Возвращает путь к лучшему checkpoint.
    """
    output_dir = f"{base_output_dir}_missing{int(missing_rate * 100):03d}"
    print(f"  Fine-tuning: missing_rate = {missing_rate:.0%}")
    print(f"  Output dir : {output_dir}")

    # Датасет
    train_texts = create_training_dataset(
        train_df_balanced, feature_names, target_name,
        prompt_config, tokenizer, missing_rate=missing_rate)
    train_dataset = Dataset.from_dict({"text": train_texts})

    def tokenize_fn(examples):
        return tokenizer(examples["text"], truncation=True, max_length=512, padding=False)

    tokenized_dataset = train_dataset.map(tokenize_fn, batched=True,
                                          remove_columns=train_dataset.column_names)

    # LoRA
    lora_config = LoraConfig(
        r=16, lora_alpha=32,
        target_modules=["q_proj", "k_proj", "v_proj", "o_proj",
                        "gate_proj", "up_proj", "down_proj"],
        lora_dropout=0.05, bias="none", task_type="CAUSAL_LM"
    )
    model_lora = get_peft_model(base_model, lora_config)
    model_lora.print_trainable_parameters()

    training_args = TrainingArguments(
        output_dir=output_dir,
        num_train_epochs=num_epochs,
        per_device_train_batch_size=batch_size,
        gradient_accumulation_steps=2,
        learning_rate=2e-4,
        bf16=True,
        logging_steps=10,
        save_strategy="epoch",
        optim="adamw_torch_fused",
        warmup_steps=50,
        max_grad_norm=1.0,
        weight_decay=0.01,
        report_to="none",
        dataloader_num_workers=4,
        group_by_length=True,
    )

    trainer = Trainer(
        model=model_lora,
        args=training_args,
        train_dataset=tokenized_dataset,
        data_collator=DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)
    )

    t0 = time.time()
    trainer.train()
    print(f"Обучение завершено за {(time.time()-t0)/60:.1f} мин")

    trainer.save_model(output_dir)

    # Возвращаем базовую модель без LoRA весов (важно для следующего эксперимента)
    model_lora = model_lora.unload()
    torch.cuda.empty_cache()

    return output_dir

In [ ]:
import glob
import re
from peft import PeftModel

def evaluate_checkpoints_on_val(output_dir, val_df, feature_names, target_name,
                                 prompt_config, tokenizer, base_model, device,
                                 eval_missing_rate=0.0):

    checkpoints = sorted(
        glob.glob(f"{output_dir}/checkpoint-*"),
        key=lambda x: int(re.search(r'checkpoint-(\d+)', x).group(1))
    )
    print(f"Найдено {len(checkpoints)} checkpoints в {output_dir}")

    best_auc, best_checkpoint = 0.0, None
    for cp in checkpoints:
        model_eval = PeftModel.from_pretrained(base_model, cp).eval()

        y_true, y_prob = [], []
        for _, row in tqdm(val_df.iterrows(), total=len(val_df),
                           desc=f"Val eval {cp.split('/')[-1]}"):
            prompt = create_prompt_missing(
                row, feature_names, target_name, prompt_config, tokenizer,
                missing_rate=eval_missing_rate)
            _, prob = predict_single_with_prob(prompt, prompt_config, model_eval, tokenizer, device)
            y_true.append(row[target_name])
            y_prob.append(prob)

        y_pred = [1 if p > 0.5 else 0 for p in y_prob]
        auc = roc_auc_score(y_true, y_prob)
        print(f"  {cp.split('/')[-1]}  ROC-AUC={auc:.4f}")

        if auc > best_auc:
            best_auc = auc
            best_checkpoint = cp

        del model_eval
        torch.cuda.empty_cache()

    print(f"Лучший checkpoint: {best_checkpoint}  (ROC-AUC={best_auc:.4f})")
    return best_checkpoint

In [ ]:
def evaluate_on_test(best_checkpoint, test_df, feature_names, target_name,
                     prompt_config, tokenizer, base_model, device,
                     eval_missing_rate=0.0):
    model_finetuned = PeftModel.from_pretrained(base_model, best_checkpoint).eval()

    y_true, y_pred, y_prob = [], [], []
    t0 = time.time()

    for _, row in tqdm(test_df.iterrows(), total=len(test_df), desc="Test eval"):
        prompt = create_prompt_missing(
            row, feature_names, target_name, prompt_config, tokenizer,
            missing_rate=eval_missing_rate)
        response, prob = predict_single_with_prob(
            prompt, prompt_config, model_finetuned, tokenizer, device)
        y_true.append(row[target_name])
        y_pred.append(parse_prediction(response, prompt_config))
        y_prob.append(prob)

    elapsed = time.time() - t0
    metrics = compute_metrics(np.array(y_true), np.array(y_pred), np.array(y_prob))
    boot = bootstrap_metrics(np.array(y_true), np.array(y_pred), np.array(y_prob), n_iter=1000)

    print("Метрики:")
    for k, v in metrics.items():
        print(f"  {k}: {v:.4f}")
    print("Bootstrap (mean±std):")
    for k, v in boot.items():
        print(f"  {k}: {v}")

    del model_finetuned
    torch.cuda.empty_cache()

    return {
        "metrics": metrics,
        "bootstrap": boot,
        "time_total": elapsed,
        "time_per_sample": elapsed / len(y_true),
        "best_checkpoint": best_checkpoint,
        "eval_missing_rate": eval_missing_rate,
    }

In [ ]:
all_results = {}

for missing_rate in MISSING_RATES:
    tag = f"missing_{int(missing_rate * 100):03d}pct"
    print(f"Эксперимент: missing_rate = {missing_rate:.0%}")

    # 7.1 Fine-tuning
    output_dir = finetune_model(
        missing_rate=missing_rate,
        train_df_balanced=train_df_balanced,
        feature_names=feature_names,
        target_name=target_name,
        prompt_config=prompt_config,
        tokenizer=tokenizer,
        base_model=base_model,
        device=device,
        num_epochs=20,
        batch_size=16,
    )

    # 7.2 Выбор лучшего checkpoint по val
    best_cp = evaluate_checkpoints_on_val(
        output_dir=output_dir,
        val_df=val_df,
        feature_names=feature_names,
        target_name=target_name,
        prompt_config=prompt_config,
        tokenizer=tokenizer,
        base_model=base_model,
        device=device,
        eval_missing_rate=missing_rate,
    )

    # 7.3 Оценка на test
    result = evaluate_on_test(
        best_checkpoint=best_cp,
        test_df=test_df,
        feature_names=feature_names,
        target_name=target_name,
        prompt_config=prompt_config,
        tokenizer=tokenizer,
        base_model=base_model,
        device=device,
        eval_missing_rate=missing_rate,
    )

    all_results[tag] = result

Эксперимент: missing_rate = 0%
  Fine-tuning: missing_rate = 0%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing000


Dataset (missing=0%):   0%|          | 0/684 [00:00<?, ?it/s]

Map:   0%|          | 0/684 [00:00<?, ? examples/s]

trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss
10,2.724837
20,1.320081
30,0.291740
40,0.140454
50,0.125457
60,0.108330
70,0.097548
80,0.095810
90,0.094433
100,0.093758


Обучение завершено за 6.8 мин
Найдено 20 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing000


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-22:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-22  ROC-AUC=0.7836


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-44:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-44  ROC-AUC=0.7641


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-66:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-66  ROC-AUC=0.7459


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-88:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-88  ROC-AUC=0.7655


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-110:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-110  ROC-AUC=0.7651


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-132:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-132  ROC-AUC=0.7700


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-154:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-154  ROC-AUC=0.7723


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-176:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-176  ROC-AUC=0.7133


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-198:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-198  ROC-AUC=0.6088


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-220:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-220  ROC-AUC=0.6405


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-242:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-242  ROC-AUC=0.6352


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-264:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-264  ROC-AUC=0.5976


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-286:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-286  ROC-AUC=0.5358


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-308:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-308  ROC-AUC=0.6195


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-330:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-330  ROC-AUC=0.5675


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-352:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-352  ROC-AUC=0.5662


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-374:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-374  ROC-AUC=0.5742


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-396:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-396  ROC-AUC=0.5680


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-418:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-418  ROC-AUC=0.5580


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-440:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-440  ROC-AUC=0.5503
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing000/checkpoint-22  (ROC-AUC=0.7836)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/150 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.7717
  F1: 0.5979
  Accuracy: 0.7400
  Precision: 0.4754
  Recall: 0.8056
Bootstrap (mean±std):
  ROC-AUC: 0.7736±0.0476
  F1: 0.5938±0.0590
  Accuracy: 0.7404±0.0342
  Precision: 0.4738±0.0638
  Recall: 0.8041±0.0665
Эксперимент: missing_rate = 20%
  Fine-tuning: missing_rate = 20%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing020


Dataset (missing=20%):   0%|          | 0/684 [00:00<?, ?it/s]

Map:   0%|          | 0/684 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/peft/mapping_func.py:72: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss
10,2.721398
20,1.310518
30,0.291567
40,0.136097
50,0.121379
60,0.113597
70,0.098545
80,0.097209
90,0.094618
100,0.093862


Обучение завершено за 6.5 мин
Найдено 20 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing020


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-22:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-22  ROC-AUC=0.7697


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-44:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-44  ROC-AUC=0.6880


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-66:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-66  ROC-AUC=0.7707


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-88:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-88  ROC-AUC=0.7679


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-110:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-110  ROC-AUC=0.7746


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-132:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-132  ROC-AUC=0.7651


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-154:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-154  ROC-AUC=0.7076


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-176:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-176  ROC-AUC=0.6518


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-198:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-198  ROC-AUC=0.6335


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-220:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-220  ROC-AUC=0.5852


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-242:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-242  ROC-AUC=0.6284


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-264:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-264  ROC-AUC=0.5870


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-286:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-286  ROC-AUC=0.5222


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-308:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-308  ROC-AUC=0.5462


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-330:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-330  ROC-AUC=0.5274


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-352:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-352  ROC-AUC=0.4944


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-374:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-374  ROC-AUC=0.5174


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-396:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-396  ROC-AUC=0.5034


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-418:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-418  ROC-AUC=0.5099


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-440:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-440  ROC-AUC=0.5115
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing020/checkpoint-110  (ROC-AUC=0.7746)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/150 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.7803
  F1: 0.6067
  Accuracy: 0.7667
  Precision: 0.5094
  Recall: 0.7500
Bootstrap (mean±std):
  ROC-AUC: 0.7822±0.0483
  F1: 0.6026±0.0605
  Accuracy: 0.7673±0.0329
  Precision: 0.5079±0.0675
  Recall: 0.7490±0.0726
Эксперимент: missing_rate = 50%
  Fine-tuning: missing_rate = 50%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing050


Dataset (missing=50%):   0%|          | 0/684 [00:00<?, ?it/s]

Map:   0%|          | 0/684 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/peft/mapping_func.py:72: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss
10,3.420837
20,1.678136
30,0.386948
40,0.154196
50,0.137448
60,0.132570
70,0.121203
80,0.118076
90,0.117317
100,0.112273


Обучение завершено за 6.4 мин
Найдено 20 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing050


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-22:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-22  ROC-AUC=0.6001


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-44:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-44  ROC-AUC=0.5489


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-66:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-66  ROC-AUC=0.6367


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-88:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-88  ROC-AUC=0.6941


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-110:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-110  ROC-AUC=0.7096


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-132:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-132  ROC-AUC=0.7548


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-154:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-154  ROC-AUC=0.6886


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-176:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-176  ROC-AUC=0.6745


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-198:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-198  ROC-AUC=0.7328


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-220:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-220  ROC-AUC=0.7182


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-242:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-242  ROC-AUC=0.7343


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-264:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-264  ROC-AUC=0.6820


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-286:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-286  ROC-AUC=0.6636


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-308:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-308  ROC-AUC=0.6374


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-330:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-330  ROC-AUC=0.6970


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-352:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-352  ROC-AUC=0.6173


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-374:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-374  ROC-AUC=0.6592


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-396:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-396  ROC-AUC=0.6612


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-418:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-418  ROC-AUC=0.6564


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-440:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-440  ROC-AUC=0.6776
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing050/checkpoint-132  (ROC-AUC=0.7548)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/150 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.6876
  F1: 0.4762
  Accuracy: 0.7067
  Precision: 0.4167
  Recall: 0.5556
Bootstrap (mean±std):
  ROC-AUC: 0.6882±0.0508
  F1: 0.4713±0.0687
  Accuracy: 0.7066±0.0373
  Precision: 0.4139±0.0715
  Recall: 0.5550±0.0849
Эксперимент: missing_rate = 90%
  Fine-tuning: missing_rate = 90%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing090


Dataset (missing=90%):   0%|          | 0/684 [00:00<?, ?it/s]

Map:   0%|          | 0/684 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/peft/mapping_func.py:72: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss
10,3.590849
20,1.728846
30,0.335792
40,0.111954
50,0.097484
60,0.093340
70,0.085370
80,0.081088
90,0.082460
100,0.078503


Обучение завершено за 6.4 мин
Найдено 20 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing090


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-22:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-22  ROC-AUC=0.4795


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-44:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-44  ROC-AUC=0.5088


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-66:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-66  ROC-AUC=0.5864


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-88:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-88  ROC-AUC=0.5743


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-110:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-110  ROC-AUC=0.6645


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-132:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-132  ROC-AUC=0.6110


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-154:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-154  ROC-AUC=0.6617


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-176:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-176  ROC-AUC=0.6746


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-198:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-198  ROC-AUC=0.6254


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-220:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-220  ROC-AUC=0.6050


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-242:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-242  ROC-AUC=0.6728


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-264:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-264  ROC-AUC=0.7413


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-286:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-286  ROC-AUC=0.5680


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-308:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-308  ROC-AUC=0.6556


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-330:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-330  ROC-AUC=0.7159


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-352:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-352  ROC-AUC=0.7044


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-374:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-374  ROC-AUC=0.6277


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-396:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-396  ROC-AUC=0.7435


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-418:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-418  ROC-AUC=0.6401


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-440:   0%|          | 0/150 [00:00<?, ?it/s]

  checkpoint-440  ROC-AUC=0.6067
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Blood_missing090/checkpoint-396  (ROC-AUC=0.7435)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/150 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.6389
  F1: 0.4270
  Accuracy: 0.6600
  Precision: 0.3585
  Recall: 0.5278
Bootstrap (mean±std):
  ROC-AUC: 0.6423±0.0526
  F1: 0.4270±0.0645
  Accuracy: 0.6626±0.0375
  Precision: 0.3598±0.0645
  Recall: 0.5332±0.0838


In [ ]:
all_results

{'missing_000pct': {'metrics': {'ROC-AUC': np.float64(0.7716861598440545),
   'F1': 0.5979381443298969,
   'Accuracy': 0.74,
   'Precision': 0.47540983606557374,
   'Recall': 0.8055555555555556},
  'bootstrap': {'ROC-AUC': '0.7736±0.0476',
   'F1': '0.5938±0.0590',
   'Accuracy': '0.7404±0.0342',
   'Precision': '0.4738±0.0638',
   'Recall': '0.8041±0.0665'},
  'time_total': 33.69546413421631,
  'time_per_sample': 0.22463642756144206,
  'best_checkpoint': '/content/drive/MyDrive/finetuned_qwen_missing_Blood_missing000/checkpoint-22',
  'eval_missing_rate': 0.0},
 'missing_020pct': {'metrics': {'ROC-AUC': np.float64(0.7803362573099415),
   'F1': 0.6067415730337079,
   'Accuracy': 0.7666666666666667,
   'Precision': 0.5094339622641509,
   'Recall': 0.75},
  'bootstrap': {'ROC-AUC': '0.7822±0.0483',
   'F1': '0.6026±0.0605',
   'Accuracy': '0.7673±0.0329',
   'Precision': '0.5079±0.0675',
   'Recall': '0.7490±0.0726'},
  'time_total': 34.004539251327515,
  'time_per_sample': 0.22669692834

In [ ]:
print("\nBootstrap (mean±std):")
for tag, res in all_results.items():
    rate = tag.replace("missing_", "").replace("pct", "%").lstrip("0") or "0%"
    print(f"\n  missing={rate}")
    for k, v in res["bootstrap"].items():
        print(f"    {k}: {v}")


Bootstrap (mean±std):

  missing=%
    ROC-AUC: 0.7736±0.0476
    F1: 0.5938±0.0590
    Accuracy: 0.7404±0.0342
    Precision: 0.4738±0.0638
    Recall: 0.8041±0.0665

  missing=20%
    ROC-AUC: 0.7822±0.0483
    F1: 0.6026±0.0605
    Accuracy: 0.7673±0.0329
    Precision: 0.5079±0.0675
    Recall: 0.7490±0.0726

  missing=50%
    ROC-AUC: 0.6882±0.0508
    F1: 0.4713±0.0687
    Accuracy: 0.7066±0.0373
    Precision: 0.4139±0.0715
    Recall: 0.5550±0.0849

  missing=90%
    ROC-AUC: 0.6423±0.0526
    F1: 0.4270±0.0645
    Accuracy: 0.6626±0.0375
    Precision: 0.3598±0.0645
    Recall: 0.5332±0.0838


ROC-AUC (0.77): Наблюдается стабильный рост по сравнению с базовыми моделями. При 20% пропусков метрика достигает пика (0.78), что доказывает высокую устойчивость модели к шуму в медицинских данных.

Recall (0.80): Высокий показатель полноты. Модель идентифицирует 80% потенциальных доноров, что критично для охвата базы. При 90% пропусков падает до 0.53 из-за потери временных признаков.

Precision (0.47): Умеренная точность из-за «агрессивной» стратегии поиска доноров. При 20% пропусков точность растет до 0.50, демонстрируя лучшую фильтрацию ложных срабатываний.

Accuracy (0.74): Стабильный уровень общей точности. Максимальное значение (0.76) зафиксировано при 20% пропусков, что подтверждает эффективность мультизадачного дообучения для неполных данных.

Время эксперимента: ~ 1 ч 14 мин.

Использовано оперативная памяти графического процесса: 45/80GB.

Графический процессор A100 80GB.